types/src/checksummed_hex.rs

```diff
@@ -0,0 +1,584 @@
+use alloc::{borrow::Cow, string::String, vec::Vec};
+use base16;
+use core::{convert::TryFrom, fmt, marker::PhantomData};
+
+use blake2::{Blake2b, Digest};
+use serde::{
+    de::{Error as DeError, Unexpected, Visitor},
+    Deserializer, Serializer,
+};
+
+/// The number of input bytes, at or below which [`encode`] and [`encode_iter`] will checksum-encode
+/// the output.
+pub const SMALL_BYTES_COUNT: usize = 75;
+
+const HEX_CHARS: [char; 22] = [
+    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f', 'A', 'B', 'C',
+    'D', 'E', 'F',
+];
+
+/// Takes a slice of bytes and breaks it up into a vector of *nibbles* (ie, 4-bit values)
+/// represented as `u8`s.
+fn bytes_to_nibbles(input: &(impl AsRef<[u8]> + ?Sized)) -> impl Iterator<Item = u8> + '_ {
+    input
+        .as_ref()
+        .iter()
+        .flat_map(move |byte| [4, 0].iter().map(move |offset| (byte >> offset) & 0x0f))
+}
+
+/// Takes a slice of bytes and outputs an infinite cyclic stream of bits for those bytes.
+fn bytes_to_bits_cycle(bytes: Vec<u8>) -> impl Iterator<Item = bool> {
+    bytes
+        .into_iter()
+        .cycle()
+        .flat_map(move |byte| (0..8usize).map(move |offset| ((byte >> offset) & 0x01) == 0x01))
+}
+
+/// Computes a Blake2b hash.
+fn blake2b_hash(data: impl AsRef<[u8]>) -> Vec<u8> {
+    let mut hasher = Blake2b::new();
+    hasher.update(data);
+    hasher.finalize().to_vec()
+}
+
+/// Encodes bytes as hexadecimal with mixed-case based checksums following a scheme similar to
+/// [EIP-55][1].
+///
+/// Key differences:
+///   - Works on any length of data, not just 20-byte addresses
+///   - Uses Blake2b hashes rather than Keccak
+///   - Uses hash bits rather than nibbles
+///
+/// **Note:** mixed-case checksummed hex will be output only if the length is <=
+/// [`SMALL_BYTES_COUNT`], otherwise, the output hex will be lowercase.
+///
+/// [1]: https://eips.ethereum.org/EIPS/eip-55
+pub fn encode(input: &(impl AsRef<[u8]> + ?Sized)) -> String {
+    encode_iter(input).collect()
+}
+
+/// `encode` but it returns an iterator.
+pub fn encode_iter(input: &(impl AsRef<[u8]> + ?Sized)) -> impl Iterator<Item = char> + '_ {
+    let input_bytes = input.as_ref();
+    let nibbles = bytes_to_nibbles(input_bytes);
+    let mut maybe_hash_bits =
+        (input_bytes.len() <= SMALL_BYTES_COUNT).then(|| bytes_to_bits_cycle(blake2b_hash(input)));
+    nibbles.map(move |mut nibble| {
+        if let Some(hash_bits) = maybe_hash_bits.as_mut() {
+            if nibble >= 10 && hash_bits.next().unwrap_or(true) {
+                nibble += 6;
+            }
+        }
+        HEX_CHARS[nibble as usize]
+    })
+}
+
+/// Returns true if all chars in a string are uppercase or lowercase.
+/// Returns false if the string is mixed case or if there are no alphabetic chars.
+fn string_is_same_case<T: AsRef<[u8]> + ?Sized>(s: &T) -> bool {
+    let mut chars = s.as_ref().iter().filter(|c| c.is_ascii_alphabetic());
```

Comment:
cheaper to check if member of `[ 'A', 'B', 'C', 'D', 'E', 'F', 'a', 'b', 'c', 'd', 'e', 'f']` I think.

Reply:
We can cast char to an integer and check if it's in expected range (`A-F` or `a-f`).

Reply:
Looks like it's already a `u8`...

Reply:
```suggestion
    let mut chars = s.as_ref().iter().filter(|c| c > 65); // 65 is the ascii code for 'A', '0'-'9' is 48-57
```

Reply:
@xcthulhu I tried this but it breaks in our proptests. I went with @goral09's suggestion.